core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java

```diff
@@ -72,6 +95,58 @@ public FilteredQueryBuilder queryName(String queryName) {
         return this;
     }
 
+    /** Returns the name of this query. */
+    public String queryName() {
+        return queryName;
+    }
+
+    @Override
+    protected boolean doEquals(FilteredQueryBuilder other) {
+        return Objects.equals(queryBuilder, other.queryBuilder) &&
+                Objects.equals(filterBuilder, other.filterBuilder) &&
+                Objects.equals(boost, other.boost) &&
+                Objects.equals(queryName, other.queryName);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(queryBuilder, filterBuilder, boost, queryName);
+    }
+
+    @Override
+    public Query toQuery(QueryParseContext parseContext) throws QueryParsingException, IOException {
+        // parsed internally, but returned null during parsing...
+        if (queryBuilder == null) {
+            return null;
+        }
+
+        if (filterBuilder == null) {
+            return queryBuilder.toQuery(parseContext);
+        }
+
+        Query query = queryBuilder.toQuery(parseContext);
```

Comment:
I think we need additional null checks here. Before the refactoring we only needed to check lucene queries in parse() method for null. Now also their `toQuery()` can potentially return `null` (e.g. if this FilteredQueryBuilder would be nested in itself). Same goes for the filter.